algorithms/src/polycommit/kzg10/mod.rs

```diff
@@ -233,10 +234,80 @@ impl<E: PairingEngine> KZG10<E> {
             hiding_bound,
         ));
 
-        let (num_leading_zeros, plain_coeffs) = skip_leading_zeros_and_convert_to_bigints(polynomial);
+        let mut commitment = match polynomial {
+            DenseOrSparsePolynomial::DPolynomial(polynomial) => {
+                let (num_leading_zeros, plain_coeffs) = skip_leading_zeros_and_convert_to_bigints(polynomial);
+
+                let msm_time = start_timer!(|| "MSM to compute commitment to plaintext poly");
+                let commitment =
+                    VariableBaseMSM::multi_scalar_mul(&powers.powers_of_beta_g[num_leading_zeros..], &plain_coeffs);
+                end_timer!(msm_time);
+
+                if terminator.load(Ordering::Relaxed) {
+                    return Err(PCError::Terminated);
+                }
+                commitment
+            }
+            DenseOrSparsePolynomial::SPolynomial(polynomial) => polynomial
+                .coeffs
+                .iter()
+                .map(|(i, coeff)| {
+                    powers.powers_of_beta_g[*i].mul_bits(BitIteratorBE::new_without_leading_zeros(coeff.to_repr()))
+                })
+                .sum(),
+        };
+
+        let mut randomness = Randomness::empty();
+        if let Some(hiding_degree) = hiding_bound {
+            let mut rng = rng.ok_or(PCError::MissingRng)?;
+            let sample_random_poly_time =
+                start_timer!(|| format!("Sampling a random polynomial of degree {}", hiding_degree));
+
+            randomness = Randomness::rand(hiding_degree, false, &mut rng);
+            Self::check_hiding_bound(
+                randomness.blinding_polynomial.degree(),
+                powers.powers_of_beta_times_gamma_g.len(),
+            )?;
+            end_timer!(sample_random_poly_time);
+        }
+
+        let random_ints = convert_to_bigints(&randomness.blinding_polynomial.coeffs);
+        let msm_time = start_timer!(|| "MSM to compute commitment to random poly");
+        let random_commitment =
+            VariableBaseMSM::multi_scalar_mul(&powers.powers_of_beta_times_gamma_g, random_ints.as_slice())
+                .into_affine();
+        end_timer!(msm_time);
```

Comment:
It seems if `hiding_bound.is_none()`, then L277 could be a cached/once_celled evaluation, which could improve performance?

How frequently do we expect `hiding_bound.is_none() == true`?

Reply:
Actually, I guess the MSM wouldn't truly evaluate any computations in that case, is that correct?

Reply:
Yeah, in the case where we don't have hiding, the randomness vector is empty, so the MSM is empty too.